crates/cargo-contract/src/cmd/extrinsics/upload.rs

```diff
@@ -126,7 +132,13 @@ impl UploadCommand {
                 }
                 Ok(())
             } else {
-                Err("This contract has already been uploaded".into())
+                let code_hash = contract_metadata::ContractMetadata::load(
+                    &crate_metadata.metadata_path(),
+                )?.source.hash;
```

Comment:
An simpler alternative to loading the metadata just to get the source hash would be to just hash the contract code inline here like https://github.com/paritytech/cargo-contract/blob/master/crates/cargo-contract/src/cmd/metadata.rs#L233.

Then would just be `blake2_hash(&code)`.

Reply:
Thanks for the suggestion. `code` is already moved though in line 122. I can of course clone it and do as you suggest, if that's what you prefer. I'm not a fan of the approach since it would duplicate the logic that takes the contract and produces the code hash and *IF* it ever changes then this would have to be updated as well. I understand that reading from the file system is probably slower but the slowness here (in the one-shot command) seems justifiable if in return we get cleaner code and more future-proof.

Reply:
Thanks for the suggestion. `code` is already moved though in line 122. I can of course clone it and do as you suggest, if that's what you prefer. I'm not a fan of the approach since it would duplicate the logic that takes the contract and produces the code hash and *IF* it ever changes then this would have to be updated as well. I understand that reading from the file system is probably slower but the slowness here (in the one-shot command) seems justifiable if in return we get cleaner code and more future-proof.

I could just do a copy of `code_hash`, right after loading `contract_metadata` in line 76, and use it here instead.

Reply:
I went with just making a copy of `code_hash` and reusing that instead. Done in [`5731420` (#804)](https://github.com/paritytech/cargo-contract/pull/804/commits/5731420b651153f1b074c63b35edf4c9feb038b5).